File: lib/services/maddeler_service.dart
import 'package:flutter/services.dart';
import 'package:csv/csv.dart';
import '../models/madde_data.dart';
import '../models/harcama_grubu_data.dart'; // AnaGrupData için
import 'github_csv_service.dart';

class MaddelerService {
  // Ana grupları gruplar.csv'den okuyacağız
  static Future<List<AnaGrupData>> loadAnaGruplar() async {
    try {
      final String csvData =
          await GitHubCSVService.loadCSVFromGitHub('gruplar.csv');
      print('gruplar.csv veri uzunluğu: ${csvData.length}');

      // Manuel olarak satırlara böl (line ending sorunları için)
      List<String> lines = csvData.split(RegExp(r'\r?\n'));
      print('Toplam satır sayısı: ${lines.length}');

      List<AnaGrupData> anaGruplar = [];
      for (int i = 1; i < lines.length; i++) {
        // İlk satır header
        String line = lines[i].trim();
        if (line.isNotEmpty) {
          List<List<dynamic>> parsedLine =
              const CsvToListConverter().convert(line);
          if (parsedLine.isNotEmpty && parsedLine[0].isNotEmpty) {
            List<dynamic> row = parsedLine[0];
            String anaGrupAdi = row[0].toString().trim();
            if (anaGrupAdi.isNotEmpty) {
              anaGruplar.add(AnaGrupData(name: anaGrupAdi));
            }
          }
        }
      }

      print('Yüklenen ana gruplar: ${anaGruplar.map((e) => e.name).toList()}');
      return anaGruplar;
    } catch (e) {
      print('Ana gruplar yükleme hatası: $e');
      throw Exception('Ana gruplar yükleme hatası: $e');
    }
  }

  // Maddeleraylik.csv'den mevcut tarihleri al
  static Future<List<String>> getAvailableDates() async {
    try {
      final String csvData =
          await GitHubCSVService.loadCSVFromGitHub('maddeleraylik.csv');
      print('maddeleraylik.csv veri uzunluğu: ${csvData.length}');

      // Manuel olarak satırlara böl
      List<String> lines = csvData.split(RegExp(r'\r?\n'));
      print('maddeleraylik.csv toplam satır sayısı: ${lines.length}');

      if (lines.isEmpty) {
        throw Exception('maddeleraylik.csv dosyası boş');
      }

      // İlk satır header - tarih sütunları 2. sütundan başlar
      String headerLine = lines[0];
      List<List<dynamic>> parsedHeader =
          const CsvToListConverter().convert(headerLine);
      List<dynamic> headerRow = parsedHeader.isNotEmpty ? parsedHeader[0] : [];

      List<String> dates = [];
      // 2. sütundan başlayarak tarih sütunlarını al
      for (int i = 2; i < headerRow.length; i++) {
        String date = headerRow[i].toString().trim();
        if (date.isNotEmpty && date.contains('-')) {
          dates.add(date);
        }
      }

      print('Bulunan tarihler: $dates');
      return dates.reversed.toList(); // En yeni tarih önce
    } catch (e) {
      print('Tarihler yükleme hatası: $e');
      throw Exception('Tarihler yükleme hatası: $e');
    }
  }

  // Seçilen ana gruba ait maddeleri ürünler.csv'den okuyacağız
  static Future<List<String>> loadMaddeler(String selectedAnaGrup) async {
    try {
      final String csvData =
          await GitHubCSVService.loadCSVFromGitHub('urunler.csv');
      print('urunler.csv veri uzunluğu: ${csvData.length}');

      // Manuel olarak satırlara böl (line ending sorunları için)
      List<String> lines = csvData.split(RegExp(r'\r?\n'));
      print('urunler.csv toplam satır sayısı: ${lines.length}');

      List<List<dynamic>> rows = [];
      for (String line in lines) {
        if (line.trim().isNotEmpty) {
          List<dynamic> row = const CsvToListConverter().convert(line).first;
          rows.add(row);
        }
      }
      print('İşlenen satır sayısı: ${rows.length}');

      if (rows.isEmpty) {
        throw Exception('urunler.csv dosyası boş');
      }

      Set<String> maddeler = {};
      String selectedGroupNorm = selectedAnaGrup.trim().toLowerCase();
      print('Aranan ana grup (normalize): $selectedGroupNorm');

      // İlk satır header, ondan sonraki satırları işle
      int matchCount = 0;
      for (int i = 1; i < rows.length; i++) {
        if (rows[i].length >= 6) {
          String anaGrup =
              rows[i][5].toString().trim().toLowerCase(); // Ana Grup sütunu
          String madde =
              rows[i][1].toString().trim().toLowerCase(); // Ürün sütunu

          // İlk birkaç eşleşme için debug
          if (anaGrup == selectedGroupNorm && matchCount < 5) {
            print(
                'Eşleşen satır ${i}: Ana Grup="${rows[i][5]}", Madde="${rows[i][1]}"');
            matchCount++;
          }

          if (anaGrup == selectedGroupNorm && madde.isNotEmpty) {
            maddeler.add(madde);
          }
        }
      }

      print('Bulunan maddeler sayısı: ${maddeler.length}');
      print('Maddeler listesi: ${maddeler.toList()}');

      return maddeler.toList();
    } catch (e) {
      print('Maddeler yükleme hatası: $e');
      throw Exception('Maddeler yükleme hatası: $e');
    }
  }

  // Maddelerin değişim oranlarını maddeleraylik.csv'den okuyacağız
  static Future<List<MaddeData>> loadMaddeDegiisimOranlari(
      List<String> maddeler, String selectedDate) async {
    try {
      final String csvData =
          await GitHubCSVService.loadCSVFromGitHub('maddeleraylik.csv');

      // Manuel olarak satırlara böl
      List<String> lines = csvData.split(RegExp(r'\r?\n'));
      print('maddeleraylik.csv satır sayısı: ${lines.length}');

      if (lines.isEmpty) {
        throw Exception('maddeleraylik.csv dosyası boş');
      }

      // Header satırından tarih sütununu bul
      String headerLine = lines[0];
      List<List<dynamic>> parsedHeader =
          const CsvToListConverter().convert(headerLine);
      List<dynamic> headerRow = parsedHeader.isNotEmpty ? parsedHeader[0] : [];

      int dateColumnIndex = -1;
      for (int i = 0; i < headerRow.length; i++) {
        if (headerRow[i].toString().trim() == selectedDate) {
          dateColumnIndex = i;
          break;
        }
      }

      if (dateColumnIndex == -1) {
        throw Exception('Seçilen tarih bulunamadı: $selectedDate');
      }

      print('Tarih sütun indeksi: $dateColumnIndex');

      List<MaddeData> maddeDataList = [];

      // Veri satırlarını işle
      for (int i = 1; i < lines.length; i++) {
        String line = lines[i].trim();
        if (line.isNotEmpty) {
          List<List<dynamic>> parsedLine =
              const CsvToListConverter().convert(line);
          if (parsedLine.isNotEmpty) {
            List<dynamic> row = parsedLine[0];

            if (row.length > dateColumnIndex) {
              String rowMaddeName = row[1].toString().trim().toLowerCase();

              // Seçilen maddeler listesinde var mı kontrol et
              if (maddeler.contains(rowMaddeName)) {
                double changeRate =
                    double.tryParse(row[dateColumnIndex].toString()) ?? 0.0;

                maddeDataList.add(MaddeData(
                  maddeName: row[1].toString().trim(), // Orijinal case'i koru
                  changeRate: changeRate,
                ));
              }
            }
          }
        }
      }

      print('Bulunan madde verileri sayısı: ${maddeDataList.length}');

      // Değişim oranına göre sırala (büyükten küçüğe)
      maddeDataList.sort((a, b) => b.changeRate.compareTo(a.changeRate));

      return maddeDataList;
    } catch (e) {
      print('Madde değişim oranları yükleme hatası: $e');
      throw Exception('Madde değişim oranları yükleme hatası: $e');
    }
  }
}

File: src/services/MaddelerService.ts
import Papa from 'papaparse';
import {GitHubCSVService} from './GitHubCSVService';
import {MaddeData} from '../models/MaddeData';

export class MaddelerService {
  static async loadAnaGruplar(): Promise<string[]> {
    try {
      const csvData = await GitHubCSVService.loadCSVFromGitHub('gruplar.csv');

      const lines = csvData.split(/\r?\n/);
      const anaGruplar: string[] = [];

      for (let i = 1; i < lines.length; i++) {
        const line = lines[i].trim();
        if (line) {
          try {
            const parsed = Papa.parse(line, {
              header: false,
              skipEmptyLines: true,
            });
            if (
              parsed.data.length > 0 &&
              Array.isArray(parsed.data[0]) &&
              parsed.data[0].length > 0
            ) {
              const anaGrupAdi = parsed.data[0][0]?.toString().trim() || '';
              if (anaGrupAdi) {
                anaGruplar.push(anaGrupAdi);
              }
            }
          } catch (e) {
            // Skip parse errors
          }
        }
      }

      return anaGruplar;
    } catch (e) {
      throw new Error(`Ana gruplar yükleme hatası: ${e}`);
    }
  }

  static async getAvailableDates(): Promise<string[]> {
    try {
      const csvData = await GitHubCSVService.loadCSVFromGitHub(
        'maddeleraylik.csv'
      );

      const lines = csvData.split(/\r?\n/);
      if (lines.length === 0) {
        throw new Error('maddeleraylik.csv dosyası boş');
      }

      const parsedHeader = Papa.parse(lines[0], {
        header: false,
        skipEmptyLines: true,
      });
      const headerRow =
        parsedHeader.data.length > 0 && Array.isArray(parsedHeader.data[0])
          ? (parsedHeader.data[0] as any[])
          : [];

      const dates: string[] = [];
      for (let i = 2; i < headerRow.length; i++) {
        const date = headerRow[i]?.toString().trim() || '';
        if (date && date.includes('-')) {
          dates.push(date);
        }
      }

      return dates.reverse(); // En yeni tarih önce
    } catch (e) {
      throw new Error(`Tarihler yükleme hatası: ${e}`);
    }
  }

  static async loadMaddeler(selectedAnaGrup: string): Promise<string[]> {
    try {
      const csvData = await GitHubCSVService.loadCSVFromGitHub('urunler.csv');

      const lines = csvData.split(/\r?\n/);
      const rows: any[][] = [];

      for (const line of lines) {
        if (line.trim()) {
          try {
            const parsed = Papa.parse(line, {
              header: false,
              skipEmptyLines: true,
            });
            if (parsed.data.length > 0 && Array.isArray(parsed.data[0])) {
              rows.push(parsed.data[0] as any[]);
            }
          } catch (e) {
            // Skip parse errors
          }
        }
      }

      const maddeler = new Set<string>();
      const selectedGroupNorm = selectedAnaGrup.trim().toLowerCase();

      for (let i = 1; i < rows.length; i++) {
        if (rows[i].length >= 6) {
          const anaGrup = rows[i][5]?.toString().trim().toLowerCase() || '';
          const madde = rows[i][1]?.toString().trim().toLowerCase() || '';

          if (anaGrup === selectedGroupNorm && madde) {
            maddeler.add(madde);
          }
        }
      }

      return Array.from(maddeler);
    } catch (e) {
      throw new Error(`Maddeler yükleme hatası: ${e}`);
    }
  }

  static async loadMaddeDegiisimOranlari(
    maddeler: string[],
    selectedDate: string
  ): Promise<MaddeData[]> {
    try {
      const csvData = await GitHubCSVService.loadCSVFromGitHub(
        'maddeleraylik.csv'
      );

      const lines = csvData.split(/\r?\n/);
      if (lines.length === 0) {
        throw new Error('maddeleraylik.csv dosyası boş');
      }

      const parsedHeader = Papa.parse(lines[0], {
        header: false,
        skipEmptyLines: true,
      });
      const headerRow =
        parsedHeader.data.length > 0 && Array.isArray(parsedHeader.data[0])
          ? (parsedHeader.data[0] as any[])
          : [];

      let dateColumnIndex = -1;
      for (let i = 0; i < headerRow.length; i++) {
        if (headerRow[i]?.toString().trim() === selectedDate) {
          dateColumnIndex = i;
          break;
        }
      }

      if (dateColumnIndex === -1) {
        throw new Error(`Seçilen tarih bulunamadı: ${selectedDate}`);
      }

      const maddeDataList: MaddeData[] = [];

      for (let i = 1; i < lines.length; i++) {
        const line = lines[i].trim();
        if (line) {
          try {
            const parsed = Papa.parse(line, {
              header: false,
              skipEmptyLines: true,
            });
            if (
              parsed.data.length > 0 &&
              Array.isArray(parsed.data[0]) &&
              parsed.data[0].length > dateColumnIndex
            ) {
              const row = parsed.data[0] as any[];
              const rowMaddeName = row[1]?.toString().trim().toLowerCase() || '';

              if (maddeler.includes(rowMaddeName)) {
                const changeRate =
                  parseFloat(row[dateColumnIndex]?.toString() || '0') || 0.0;

                maddeDataList.push({
                  maddeName: row[1]?.toString().trim() || '', // Orijinal case'i koru
                  changeRate: changeRate,
                });
              }
            }
          } catch (e) {
            // Skip parse errors
          }
        }
      }

      maddeDataList.sort((a, b) => b.changeRate - a.changeRate);
      return maddeDataList;
    } catch (e) {
      throw new Error(`Madde değişim oranları yükleme hatası: ${e}`);
    }
  }

  static async getMaddeMonthlyChangeData(maddeAdi: string): Promise<{
    dates: string[];
    values: number[];
  }> {
    try {
      const csvData = await GitHubCSVService.loadCSVFromGitHub(
        'maddeleraylik.csv'
      );

      const lines = csvData.split(/\r?\n/);
      if (lines.length === 0) {
        return {dates: [], values: []};
      }

      const parsedHeader = Papa.parse(lines[0], {
        header: false,
        skipEmptyLines: true,
      });
      const headerRow =
        parsedHeader.data.length > 0 && Array.isArray(parsedHeader.data[0])
          ? (parsedHeader.data[0] as any[])
          : [];

      // Tarihleri başlık satırından al (3. kolondan itibaren)
      const dates: string[] = [];
      for (let i = 2; i < headerRow.length; i++) {
        const tarihStr = headerRow[i]?.toString().trim() || '';
        if (tarihStr && tarihStr.includes('-')) {
          try {
            const dateParts = tarihStr.split('-');
            if (dateParts.length === 3) {
              const monthName = MONTH_NAMES[dateParts[1]] || dateParts[1];
              dates.push(`${monthName} ${dateParts[0]}`);
            } else {
              dates.push(tarihStr);
            }
          } catch (e) {
            dates.push(tarihStr);
          }
        }
      }

      // Madde adını normalize et
      const normalizedMaddeAdi = maddeAdi.trim().toLowerCase();

      // Veri satırlarında maddeyi bul
      const values: number[] = [];
      for (let i = 1; i < lines.length; i++) {
        const line = lines[i].trim();
        if (!line) continue;
        try {
          const parsed = Papa.parse(line, {
            header: false,
            skipEmptyLines: true,
          });
          if (
            parsed.data.length > 0 &&
            Array.isArray(parsed.data[0]) &&
            parsed.data[0].length >= 2
          ) {
            const row = parsed.data[0] as any[];
            const rowMaddeName = row[1]?.toString().trim().toLowerCase() || '';

            if (rowMaddeName === normalizedMaddeAdi) {
              // Aylık değişim değerlerini oku (3. kolondan itibaren)
              for (let j = 2; j < row.length && j <= dates.length + 1; j++) {
                const value = parseFloat(row[j]?.toString() || '0') || 0.0;
                values.push(value);
              }
              break; // Madde bulundu, döngüden çık
            }
          }
        } catch (e) {
          // Skip parse errors
          continue;
        }
      }

      return {dates: dates, values: values};
    } catch (e) {
      console.log('Madde aylık değişim veri yükleme hatası:', e);
      return {dates: [], values: []};
    }
  }

  static async getMaddeYearlyChange(maddeAdi: string): Promise<number> {
    try {
      const csvData = await GitHubCSVService.loadCSVFromGitHub(
        'maddeleryıllık.csv',
        false // Cache'i devre dışı bırak
      );

      const lines = csvData.split(/\r?\n/);
      if (lines.length === 0) {
        return 0.0;
      }

      const parsedHeader = Papa.parse(lines[0], {
        header: false,
        skipEmptyLines: true,
      });
      const headerRow =
        parsedHeader.data.length > 0 && Array.isArray(parsedHeader.data[0])
          ? (parsedHeader.data[0] as any[])
          : [];

      // Madde adını normalize et
      const normalizedMaddeAdi = maddeAdi.trim().toLowerCase();

      // Header'dan madde kolonunu bul
      let maddeColumnIndex = -1;
      for (let i = 1; i < headerRow.length; i++) {
        const headerName = headerRow[i]?.toString().trim().toLowerCase() || '';
        if (headerName === normalizedMaddeAdi) {
          maddeColumnIndex = i;
          break;
        }
      }

      if (maddeColumnIndex === -1) {
        console.log(`Madde bulunamadı: ${maddeAdi}`);
        return 0.0;
      }

      // Son satırdan (en son tarih) değeri oku
      for (let i = lines.length - 1; i >= 1; i--) {
        const line = lines[i].trim();
        if (!line) continue;
        
        try {
          const parsed = Papa.parse(line, {
            header: false,
            skipEmptyLines: true,
          });
          if (
            parsed.data.length > 0 &&
            Array.isArray(parsed.data[0]) &&
            parsed.data[0].length > maddeColumnIndex
          ) {
            const row = parsed.data[0] as any[];
            const value = parseFloat(row[maddeColumnIndex]?.toString() || '0');
            
            // Eğer değer geçerli bir sayıysa döndür
            if (!isNaN(value) && isFinite(value)) {
              return value;
            }
          }
        } catch (e) {
          // Skip parse errors
          continue;
        }
      }

      return 0.0;
    } catch (e) {
      console.log('Madde yıllık değişim okuma hatası:', e);
      return 0.0;
    }
  }
}

const MONTH_NAMES: {[key: string]: string} = {
  '01': 'Oca',
  '02': 'Şub',
  '03': 'Mar',
  '04': 'Nis',
  '05': 'May',
  '06': 'Haz',
  '07': 'Tem',
  '08': 'Ağu',
  '09': 'Eyl',
  '10': 'Eki',
  '11': 'Kas',
  '12': 'Ara',
};

